packages/app/src/app/overmind/internalActions.ts

```diff
@@ -581,62 +582,68 @@ export const setViewModeForDashboard = ({ effects, state }: Context) => {
   effects,
 }: Context) => {
-  const { id, isValid } = await actions.internal.getTeamIdFromUrlOrStore();
+  const persistedWorkspaceId = actions.internal.getTeamIdFromUrlOrStore();
+
+  const hasWorkspaces =
+    state.dashboard?.teams && state.dashboard.teams.length > 0;
+  if (!hasWorkspaces) {
+    const teams = await effects.gql.queries.getTeams({});
 
-  if (isValid && id) {
+    if (teams?.me) {
+      state.dashboard.teams = teams.me.workspaces;
+      state.primaryWorkspaceId = teams.me.primaryWorkspaceId;
+      state.userCanStartTrial = teams.me.eligibleForTrial;
+    }
+
+    // TODO: Treat here future scenario when no workspaces are available
+  }
+
+  const isPersistedWorkspaceValid = state.dashboard.teams.some(
+    team => team.id === persistedWorkspaceId
+  );
+
+  if (isPersistedWorkspaceValid && persistedWorkspaceId) {
     // Set active team from url or storage.
-    actions.setActiveTeam({ id });
+    actions.setActiveTeam({ id: persistedWorkspaceId });
   } else {
-    // Change to first pro workspace or personal space if no pro exists
-    // TODO: In the future, the fallback will be the first workspace in
-    // the list, when no personal exists
-    actions.internal.inferWorkspaceFromUserData();
+    actions.internal.setFallbackWorkspace();
   }
 };
 
-export const getTeamIdFromUrlOrStore = async ({
-  state,
-  effects,
-  actions,
-}: Context): Promise<{ id: string | null; isValid: boolean }> => {
-  const suggestedTeamId =
-    actions.internal.getTeamIdFromUrl() ||
-    actions.internal.getTeamIdFromLocalStorage();
-  const hasTeams = state.dashboard?.teams && state.dashboard.teams.length > 0;
-
-  let userTeams: TeamFragmentDashboardFragment[] | undefined;
-
-  if (hasTeams) {
-    userTeams = state.dashboard.teams;
+export const setFallbackWorkspace = ({ actions, state }: Context) => {
+  if (state.primaryWorkspaceId) {
+    actions.setActiveTeam({ id: state.primaryWorkspaceId });
   } else {
-    // TODO: Instead of actions.dashboard.getTeams();
-    // We might be able to use it though, and then use state.dashboard.teams!
-    const teams = await effects.gql.queries.getTeams({});
-
-    if (teams?.me) {
-      userTeams = teams.me.workspaces;
+    const firstWorkspace =
+      state.dashboard.teams.length > 0 ? state.dashboard.teams[0] : null;
+    const firstProWorkspace = state.dashboard.teams.find(
+      team =>
+        team.subscription?.status === SubscriptionStatus.Active ||
+        team.subscription?.status === SubscriptionStatus.Trialing
+    );
 
-      // Also set state while we're at it
-      state.dashboard.teams = teams.me.workspaces;
-      state.personalWorkspaceId = teams.me.personalWorkspaceId;
-      state.userCanStartTrial = teams.me.eligibleForTrial;
+    if (firstProWorkspace) {
+      actions.setActiveTeam({ id: firstProWorkspace.id });
+    } else if (firstWorkspace) {
+      actions.setActiveTeam({ id: firstWorkspace.id });
+    } else {
+      // TODO: Redirect to workspace setup
```

Comment:
Is this an uncomplete step?

Reply:
Yes, this will be implemented in https://linear.app/codesandbox/issue/PC-1341/handle-empty-state-for-primary-workspaces, will drop the link there as well